lisa/tools/lagscope.py

```diff
@@ -0,0 +1,254 @@
+from decimal import Decimal
+from typing import List, Type, cast
+
+from lisa.executable import Tool
+from lisa.messages import NetworkLatencyPerformanceMessage
+from lisa.operating_system import Debian, Posix, Redhat, Suse
+from lisa.util import LisaException, find_patterns_groups_in_lines, get_datetime_path
+
+from .git import Git
+
+
+class Lagscope(Tool):
+    repo = "https://github.com/Microsoft/lagscope"
+    # lagscope 1.0.1
+    # ---------------------------------------------------------
+    # 02:06:15 INFO: New connection: local:25001 [socket:3] --> 10.0.0.5:6001
+    # 02:11:12 INFO: TEST COMPLETED.
+    # 02:11:12 INFO: Ping statistics for 10.0.0.5:
+    # 02:11:12 INFO:    Number of successful Pings: 1000000
+    # 02:11:12 INFO:    Minimum = 135.500us, Maximum = 9644.250us, Average = 294.313us
+    # 02:11:12 INFO: Dumping all latencies into csv file: Latency-20220106-0817.csv
+    #
+    # Percentile       Latency(us)
+    #      50%         295
+    #      75%         345
+    #      90%         362
+    #      95%         376
+    #      99%         404
+    #    99.9%         620
+    #   99.99%         135
+    #  99.999%         2654
+    _result_pattern = re.compile(
+        r"([\w\W]*?)Minimum = (?P<min_latency_us>.+?)us, "
+        r"Maximum = (?P<max_latency_us>.+?)us, Average = (?P<average_latency_us>.+?)us"
+        r"([\w\W]*?)95%\s+(?P<latency95_percentile_us>\d+)"
+        r"([\w\W]*?)99%\s+(?P<latency99_percentile_us>\d+)",
+        re.M,
+    )
+    # Interval(usec)   Frequency
+    #       0          0
+    #      30          0
+    #      45          0
+    #      60          0
+    #      75          0
+    #      90          0
+    #     105          0
+    #     120          0
+    #     135          1013
+    #     150          7111
+    #     165          15318
+    #     180          35440
+    #     195          39818
+    #     210          63977
+    #     225          51341
+    #     240          62950
+    #     255          78099
+    #     270          88724
+    #     285          76162
+    #     300          52777
+    #     315          76238
+    #     330          93312
+    #     345          141943
+    #     360          62452
+    #     375          32662
+    #     390          10980
+    #     405          3605
+    #     420          1786
+    #     435          992
+    #     450          639
+    #     465          457
+    #     480          2204
+    _interval_frequency_pattern = re.compile(
+        r"\s+(?P<interval_us>\d+)\s+(?P<frequency>\d+)$", re.M
+    )
+
+    @property
+    def dependencies(self) -> List[Type[Tool]]:
+        return [Git]
+
+    @property
+    def command(self) -> str:
+        return "lagscope"
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _install(self) -> bool:
+        self._install_dep_packages()
+        tool_path = self.get_tool_path()
+        git = self.node.tools[Git]
+        git.clone(self.repo, tool_path)
```

Comment:
Choose a release tag or branch to make the test case more stable. It's similar to other source installed tools.

Reply:
It can be set in the clone method. It's simpler.

Reply:
updated